Add foreman_templates manual

ares opened the pull request.

ares committed:
Add foreman_templates manual

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+
+# 1. {{ page.title }}
+
+This plugin enables synchronization of provisioning templates, partition tables, job templates from external git repository and/or file system. It can be used both with the [community templates](https://github.com/theforeman/community-templates) repository or with any other source of templates that has the same directory structure. Please note that Foreman core provides tracking of changes, brows ...
```
"*and* job templates" will finish the list correctly, otherwise it's dangling

sean797 replied:
Does it require the same directory structure? I thought it worked with the metadata and all files needed to be suffixed with `.erb`

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+
+This plugin enables synchronization of provisioning templates, partition tables, job templates from external git repository and/or file system. It can be used both with the [community templates](https://github.com/theforeman/community-templates) repository or with any other source of templates that has the same directory structure. Please note that Foreman core provides tracking of changes, brows ...
+
+You can use the native git support for simple syncrhonization with remote git repositories, which is usually useful for syncing new template versions from community repository or for exporting local changes to contribute back. Or if you have more complicated workflow, you can point the plugin to work with local directory. This way, you can use any VCS (including git) to control the versioning of  ...
```
I think "use this plugin to sync the directory when required." might read better here, but it's minor.

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+    foreman-installer --enable-foreman-plugin-templates
+
+If installed foreman from git, you can simply add foreman-templates gem as one of the dependencies. Don't forget to run migrations and seed afterwards.
+      
```
No mention of install from packages here? IMO it's preferable to using the installer, but I know I'm in a minority :)

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+  </tr>
+  <tr>
+    <td><i>branch</i></td>
+    <td colspan="2">git branch to read or write from/to (only for git url), default value is based on current Foreman version - e.g. develop or 1.15-stable</td>
```
I'd say "only for git-based repositories" in the parentheses, and maybe put the "repo" option immediately before this one so it's easily referenced.

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+
+After the installation there are new settings available. These are the default setting for both importing and exporting tasks. They can be overridden on each synchronization run and serve only as default so users don't have to specify the same value on every run. You will find them under Administer -> Settings -> TemplateSync. Only administrator can modify their values. Following table explains t ...
+
+<table class="table table-bordered table-condensed table-striped">
```
I see you're going for alphabetical setting order - but many options are rarely used, does it makes sense to put common options first?

ares replied:
This is also the order we have in settings table. It's not a huge number so I'd keep it ordered alphabetically

GregSutcliffe replied:
We'll have to agree to disagree on this, as I feel for docs the user should get the most relevant info first - but I don't feel strongly enough to insist on it :)

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+
+# 4. Usage (features description)
+
+Both importing and exporting is available as rake tasks. Rake tasks are triggered from terminal. If you installed the plugin using foreman-installer, or simply it's package based installation), you should be able to run rake using `foreman-rake` command. If you installed Foreman alternatively, e.g. cloned from git, you should likely use `RAILS_ENV=production bundle exec rake` ran from your Forema ...
```
Package install mentioned here, so it should definitely go under Installation too (see earlier comment)

GregSutcliffe replied:
s/If you installed Foreman alternatively, e.g. cloned from git, you should likely use/Source-based installs of Foreman will need to use Bundler to call Rake as normal, e.g./

GregSutcliffe replied:
I'd put the API sentence on a new line/paragraph to make it stand out for those who want to know.

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
````diff
@@ -0,0 +1,185 @@
+foreman-rake templates:import repo=https://github.com/theforeman/community-templates
+```
+
+The example above would clone the repository to some temp directory. If `branch` was specified, it checkouts this git branch. Then if `dirname` was provided, it will "cd" into it. Finally is searches for all files in this directory and subdirectories that has `'.erb` suffix. Then the template is parsed and imported (see below for more details)
````
s/it checkouts this git branch/it switches to this branch/

GregSutcliffe replied:
s/Finally is/Finally it/

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,185 @@
+
+Compared to previous example, this uses local directory. Therefore not git clone or branch checkout is done. If `dirname` is specified, the task goes into this subdirectory and searches for all files with '.erb' extension. Then the same parsing and importing process follows,
+
+The parsing process reads the content of the file. It expects the content of the template including a metadata. The metadata are an ERB comment on top of the file. It's in yaml format and specifies template name, kind, compatible operating systems, organizations, locations and possibly other. The template gets it's name based on information from metadata. Before the template is saved in database, ...
```
"including a metadata block." or "including some template metadata" - your choice :)

GregSutcliffe replied:
Is it worth linking to an example from community-templates to show a metadata block?

GregSutcliffe reviewed (changes requested):
Looks pretty good to me, thanks @ares! A few small typos and wording changes, that's all.

sean797 commented on plugins/foreman_templates/4.0/index.md:
````diff
@@ -0,0 +1,185 @@
+foreman-rake templates:import
+```
+
+This task based on $repo setting either uses git or scans local directory for templates to import. If you want to customize any of setting for a particular run, just specify it as extra argument.
````
s/If you want to customize any of setting/If you want to customize any setting/

I'm not sure `of` makes sense here.

sean797 commented on plugins/foreman_templates/4.0/index.md:
````diff
@@ -0,0 +1,185 @@
+
+```sh
+# prefix all templates with "[comunity] " string to avoid overriding existing templates
+foreman-rake templates:import prefix='[community] '
````
Not required but it might be nice to show something like:

```sh
$ tree /tmp/my-repo/
/tmp/my-repo/
└── large-ptable.erb

0 directories, 1 file
$ foreman-rake templates:import repo=/tmp/my-repo/ prefix='[community] '
```
Would result in a template named `[community] large-ptable`.

ares replied:
added as a new example with a comment

sean797 commented on plugins/foreman_templates/4.0/index.md:
````diff
@@ -0,0 +1,185 @@
+foreman-rake templates:export
+```
+
+If the `repo` is git repository, we clone it to some temp directory. Then we apply changes to template files and if there was a change, we automatically create a commit with generate message and push it back to origin. We do not delete files that are not present in Foreman database, so this should not be destructive. On the other hand if you want more control over the process or you want to use a ...
````
s/Then we apply changes to template files and if there was a change, we automatically create a commit with generate message and push it back to origin./Then we apply changes to template files and automatically create a commit with a generate message and push it back to origin./

Also

s/clone it to some temp directory/clone it to a temp directory/

Also

s/To specify local directory/To specify a local directory/

ares commented:
Thanks @GregSutcliffe, I think I addressed all your comments. Please take another look.

@sean797 same for you. Just one answer instead of a text change

> Does it require the same directory structure? I thought it worked with the metadata and all files needed to be suffixed with .erb

It does not required the same structure and it works as you said for import. But export creates the same structure the community-templates has, so if you want to use both direction sync, you should use this structure.

sean797 reviewed (approved):
LGTM @ares ! Thanks for doing this 👍

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,198 @@
+
+    foreman-installer --enable-foreman-plugin-templates
+
+If you installed the plugin using package directly, you should run the migrations and seed (just to be sure) and restart the Foreman.
```
Can we either detail these steps or link to where they can be found? If this is the first plugin a user ever installs, they may not know how...

ares replied:
I would prefer not documenting it in details. foreman-installer is the recommended way. We also don't document you could do `git clone ...`

GregSutcliffe replied:
I don't think we need details (they go out of date) but maybe a link to https://theforeman.org/plugins/#2.2Packageinstallation ?

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,198 @@
+
+The result of the whole import is printed on the STDOUT. The output contains list of created/updated templates as well as applied diffs.
+
+Few more useful examples follow
```
nitpick: *A* few more ...

GregSutcliffe commented on plugins/foreman_templates/4.0/index.md:
```diff
@@ -0,0 +1,198 @@
+
+# 8. Links
+
+* foreman_templates plugin [https://github.com/theforeman/foreman_chef](https://github.com/theforeman/foreman_templates)
```
Url is right, but text says Chef :P

ares replied:
damn, I swear I grepped it :-)

GregSutcliffe commented:
@ares Almost there, a couple of tiny things :)

ares commented:
@GregSutcliffe thanks, another round? :-)

GregSutcliffe commented:
Merged, thanks @ares!
